### PR TITLE
perf: topology-grouped vmap Hessian batching in JaxLoss

### PR DESCRIPTION
**Pull request by ericchansen**
## Summary

Topology-grouped `jax.vmap` Hessian batching in `JaxLoss`. Molecules sharing the same topology (same energy function) now compute their Hessians in a single vectorized call instead of one traced call per molecule.

### Rh-enamide benchmarks (9 molecules, 2,739 params, CPU)

| Metric | Before | After | Change |
|--------|-------:|------:|-------:|
| Loss eval | 632 ms | 472 ms | **−25%** |
| JIT compile | 8.5 s | 7.3 s | **−14%** |
| Loss value | 91.4817 | 91.4817 | exact match ✅ |

### How it works

- Groups molecules by topology signature (reuses `batched._topology_signature`)
- Multi-molecule groups get a single `jax.vmap(jax.hessian(...))` call
- Singletons keep the existing per-molecule path
- Downstream freq/eigenmatrix/residual computation stays per-molecule (different reference data arrays)
- No padding, no eigendecomposition corruption — real-size matrices throughout

### Tests

- 2 new tests in `TestJaxLossTopologyBatching`:
  - Same-topology vmap path (two water molecules)
  - Mixed-topology parity vs Python `ObjectiveFunction` (H₂ + 2× water)
- All 712 existing tests pass

Refs #244.

### CODE REVIEW

File: q2mm/optimizers/jaxloss.py [hunk 1]
```diff
@@ -304,17 +311,103 @@ def _build(self) -> None:
...
+        # Group molecules that share the same topology (same handle →
+        # same energy function).  For multi-molecule groups, build a
+        # single vmapped Hessian function instead of tracing one per
+        # molecule.  This reduces JIT compilation time and can improve
+        # GPU utilisation.
+        from q2mm.backends.mm.batched import _topology_signature
+
+        hess_groups: dict[str, dict] = {}
+        entry_to_group: dict[int, str] = {}
+
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxloss.py:324]: Fixed in 8ec9b7c. Removed the unused `entry_to_group` dict.

File: q2mm/optimizers/jaxloss.py [hunk 2]
```diff
@@ -304,17 +311,103 @@ def _build(self) -> None:
...
             mol_data.append(entry)
 
+        # ---- Topology-grouped Hessian batching ----
+        #
+        # Group molecules that share the same topology (same handle →
+        # same energy function).  For multi-molecule groups, build a
+        # single vmapped Hessian function instead of tracing one per
+        # molecule.  This reduces JIT compilation time and can improve
+        # GPU utilisation.
+        from q2mm.backends.mm.batched import _topology_signature
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxloss.py:321]: Intentional divergence. `batched.py` extracts params from a `ForceField` object (`forcefield.get_param_vector()`), but inside `_loss_fn` params must be the function argument (what `jax.grad` differentiates w.r.t.). We do reuse `_topology_signature` from `batched.py` to ensure grouping logic stays in one place.

File: test/test_jax_loss.py [hunk 3]
```diff
@@ -822,3 +822,87 @@ def test_ts_mol_indices_rejects_out_of_range(self) -> None:
...
+
+        ff = ForceField(
+            bonds=[BondParam(elements=("H", "O"), force_constant=400.0, equilibrium=1.05)],
+            angles=[AngleParam(elements=("H", "O", "H"), force_constant=35.0, equilibrium=110.0)],
+        )
+
+        ref = ReferenceData()
+        ref.add_energy(value=0.0, molecule_idx=0, weight=1.0)
+        ref.add_energy(value=0.0, molecule_idx=1, weight=1.0)
+
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply test/test_jax_loss.py:859]: Fixed in 8ec9b7c. Rewrote as `test_two_water_freq_batching_parity` — now adds frequency references for both molecules so `needs_hessian_computation` is True and the vmapped Hessian path is exercised. Asserts parity against Python ObjectiveFunction.

File: test/test_jax_loss.py [hunk 4]
```diff
@@ -822,3 +822,87 @@ def test_ts_mol_indices_rejects_out_of_range(self) -> None:
...
+                AngleParam(elements=("H", "O", "H"), force_constant=49.9, equilibrium=104.5),
+            ],
+        )
+
+        # Mix energy + frequency refs across different topologies
+        ref = ReferenceData()
+        ref.add_energy(value=0.0, molecule_idx=0, weight=1.0)
+        ref.add_energy(value=0.0, molecule_idx=1, weight=1.0)
+        ref.add_energy(value=0.0, molecule_idx=2, weight=1.0)
+
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply test/test_jax_loss.py]: Fixed in 8ec9b7c. Rewrote as `test_mixed_topology_freq_parity` — adds frequency refs for all three molecules (H2 singleton group + 2x water vmapped group). Both topology-group paths are now exercised.
